### PR TITLE
Change to InputEventMIDI event to include connection_index property

### LINKED ISSUE #7733: Provide MIDI device property in InputEventMIDI class
### Describe the project you are working on

I am working on a digital art project where I am using Godot as visualization engine which is influenced by audio analysis and incoming MIDI events.

### Describe the problem or limitation you are having in your project

Some external MIDI Hardware devices cannot change their MIDI channel and cannot be distinguished from each other without device information for each incoming MIDI event.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

A string attribute "input" or "device" containing the device name as listed by OS.get_connected_midi_inputs() or an int attribute pointing to the corresponding array element of OS.get_connected_midi_inputs() would solve the problem.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

if input_event is InputEventMIDI and input_event.device == "my_awesome_midi_controller":
    do_something_useful()

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I haven't found any workaroung except integrating an external MIDI library like portmidi or rtmidi as GDExtension

### Is there a reason why this should be core and not an add-on in the asset library?

All necessary information should already be available and exposing the device name should be quiet simple (i would guess).

### PR DESCRIPTION
**Pull request by ramdor**
Added connection_index to the InputEventMIDI so that each MIDI message can be associated with a specific opened MIDI device.

- *Production edit: This closes https://github.com/godotengine/godot-proposals/issues/7733.*

### CODE REVIEW

**Comment by Calinou** [comment]
Does this close https://github.com/godotengine/godot-proposals/issues/7733?

**Comment by ramdor** [comment]
@Calinou yes. The connection index is direcly related to the count through the midi devices that are opened.

File: core/input/input_event.h [hunk 1]
```diff
@@ -547,6 +548,9 @@ class InputEventMIDI : public InputEvent {
 	void set_controller_value(const int p_controller_value);
 	int get_controller_value() const;
 
+	void set_connection_index(const int p_connection_index);
```
Comment (AThousandShips) [inline core/input/input_event.h:551]:
```suggestion
	void set_connection_index(int p_connection_index);
```
While it's used elsewhere here we don't use `const` for simple types so let's not use it for new methods

File: core/input/input_event.cpp [hunk 2]
```diff
@@ -1717,6 +1717,14 @@ int InputEventMIDI::get_controller_value() const {
 	return controller_value;
 }
 
+void InputEventMIDI::set_connection_index(const int p_connection_index) {
```
Comment (AThousandShips) [inline core/input/input_event.cpp:1720]:
```suggestion
void InputEventMIDI::set_connection_index(int p_connection_index) {
```

**Comment by akien-mga** [comment]
Shouldn't this be stored in the existing `InputEvent.device` property instead of adding a new parameter everywhere?

**Comment by ramdor** [comment]
I can make all the changes and swap it to use InputEvent.device np, removing the need for the extra parameter and fix the documentation.

**Comment by ramdor** [comment]
thanks for the review, and for the advice on using the InputEvent device instead. I have made the tweaks and submitted the changes. Cheers.

**Pull request closed without merging**